### PR TITLE
Add descriptions parameter for listing tasks in the transport layer

### PR DESCRIPTION
**Pull request by benwtrent**
This commit adds a new `descriptions` parameter for listing tasks in the transport layer. 

This allows tasks to be filtered by `actions` and `descriptions` so that matching specific tasks
within a particular action is possible. 

NOTE: This is not available as a REST parameter yet. The only way to use this parameter is through the transport client within the cluster.

related to: https://github.com/elastic/elasticsearch/issues/70008

### CODE REVIEW

**Comment by benwtrent** [comment]
//CC @elastic/ml-core

**Comment by rjernst** [comment]
It seems like we are moving more and more away from the patterns we use within other rest APIs, in that instead of having an actual request body, this API is getting more url parameters. Could we instead make this a real json request? Url parameters have, historically, been about tweaking how the request and/or response is displayed, with some caveats of any named parameter being automatically supported through a url parameter. But these parameters are clearly request behavior logic.

**Comment by benwtrent** [comment]
@rjernst if a larger refactor of the API is desired, I would prefer it done in a separate PR and preferably by the core infra team :). 

I am adding this parameter as it is a blocker for allowing system index cleanup work for machine learning and transforms.

**Comment by benwtrent** [comment]
run elasticsearch-ci/1
run elasticsearch-ci/2

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
@rjernst I could remote the REST update to this and simply add it to the backend. ML would be using the transport action directly. I added the REST change for completeness but I can remove it.

**Comment by rjernst** [comment]
I wonder if we need a new parameter for this at all. The tasks api is still marked as beta. The descriptions seem like they should simply always be output when `detailed=true`. We don't need to fix the existing request right now, but at least let's not add more parameters that don't fit our api pattern.

**Comment by benwtrent** [comment]
@rjernst its not about requesting a description, it's about filtering the requested tasks via their descriptions. 

For machine learning, we need to wait until there are no bulk indexing tasks for machine learning indices. This is for the system index cleanup work.

Honestly, I would rather not add the parameter at all (body or URL). But adding just one to the body while the rest are URL parameters is messy and confusing. So, I am all for just not adding it to the REST action for right now.

**Comment by rjernst** [comment]
Could you not filter on the ML side after getting the tasks?

**Comment by benwtrent** [comment]
@rjernst 
> Could you not filter on the ML side after getting the tasks?

The great thing about the lists tasks API is `.setWaitForCompletion(true)` and setting a timeout. That is what we use it for mostly in the stack. That is what we want to use it for here. Wait for these tasks to complete :)

**Comment by benwtrent** [comment]
@rjernst 

Either we add a single parameter to this transport method (and possibly the rest endpoint).

Or the ML team creates a whole transport action that is effectively the same but with this added parameter. 

I think those are our only two options. Since the new transport action would be a copy-paste of the List Actions, I opted for adding there instead of a whole new action.

**Comment by rjernst** [comment]
Alright, I think just adding to the request object but not the rest API would be better for now. I would like to not add anything more that we must maintain REST compatibility for until we've had time to assess this API.

**Comment by benwtrent** [comment]
run elasticsearch-ci/2

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksRequest.java [hunk 1]
```diff
@@ -8,34 +8,73 @@
...
+        if (descriptions.length > 0 && detailed == false) {
+            validationException = addValidationError("matching on descriptions is not available when [detailed] is false",
+                validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    public boolean match(Task task) {
+        if (super.match(task) == false) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksRequest.java:71]: I find this less readable than a single multi-line boolean expression:

```
return super.match(task) && 
      (isEmpty(getDescriptions()) || simpleMatch(getDescriptions(), task.getDescription()));
```

**Review by rjernst — APPROVED** [review]
One request, otherwise LGTM

**Comment by benwtrent** [comment]
@pugnascotia while this is an `>enhancement` no updates were made to user APIs. Its only accessible via the transport layer. Unsure how we can communicate that. 

Also, the docs make no mention of the parameter.

**Comment by pugnascotia** [comment]
I think I'll change it to `>non-issue` in that case.

**Pull request merged**